### PR TITLE
Adds staff and instructor users of the master course to CCX

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
### Background
fixes https://github.com/mitocw/edx-platform/issues/126, fixes https://github.com/mitocw/edx-platform/issues/155, fixes https://github.com/mitocw/edx-platform/issues/115 and fixes https://github.com/mitocw/edx-platform/issues/205

This is an updated version of https://github.com/edx/edx-platform/pull/10877 which had to be reverted. 

### What is done in this PR
**Studio Updates:** 
- Remove ccx from staff dashboard inside studio. Because  a master course can have unlimited ccx, staff's dashboard can be populated with a huge list of ccx along with master course 

**LMS Updates:** 
- This is for course staff to see what ccx coaches are doing with course.
- Now when a coach creates a new ccx, I am assigning staff role on this ccx course to all staff of master course.
- Added instructors of master course as instructor(s) in ccx.
- Also adds the master course staff users to new CCX created via the CCX REST APIs
- **Note:** This code will add/enrol staff and instructors in ccx and now max student limit for ccx depends on number of staff and admins. 
- Fixed view as student issue on ccx, now staff can view as student and see how ccx  view will appear to student
- Fixed display name on ccx coach dashboard. Previously it was show display name of master course on coach dashboard.

##### To run migration manually on devstack
- ``python manage.py lms migrate ccx 0003 --settings=devstack``
##### To reverse migration manually on devstack
- ``python manage.py lms migrate ccx 0002 --settings=devstack``

@pdpinch @giocalitri 

- Staff on Master course
screen shot 2015-11-30 at 5 12 54 pm

- Staff on CCX
screen shot 2015-11-30 at 5 13 04 pm

- Admin/instructors on master course
screen shot 2015-12-10 at 5 10 18 pm

- Admin/instructors on ccx
screen shot 2015-12-10 at 5 10 22 pm

#### Fixed view as student

screen shot 2015-12-17 at 7 21 37 pm 2
screen shot 2015-12-17 at 7 21 53 pm 2
screen shot 2015-12-17 at 7 22 05 pm 2

#### Fixed display name of ccx on coach dashboard.
screen shot 2015-12-18 at 3 29 34 pm
screen shot 2015-12-18 at 3 29 44 pm
screen shot 2015-12-18 at 3 29 55 pm

### CODE REVIEW

**Comment by pdpinch** [comment]
This looks to have all the changes I asked for, but the diff isn't correct because master doesn't include the revert from bf5d203

Can you update the cover letter to explain that this updates & restores 91bf48f which was reverted by bf5d203. Also, we need to cut the title down. It's way too long.

**Comment by pdpinch** [comment]
@giocalitri has some concerns about the reverse migration which he brought up over chat. 

```
I am now confident that the 0002 reverse migration cannot work as it is 
because during the migration all the `lazy` decorated methods of the 
CCX model cannot be found.

this is how the reverse migration should be to work:
```

```
def reverse_add_master_course_staff_to_ccx_for_existing_ccx(apps, schema_editor):
    """
    Add all staff and admin of master course to respective CCX(s).
    """
    CustomCourseForEdX = apps.get_model("ccx", "CustomCourseForEdX")
    list_ccx = CustomCourseForEdX.objects.all()
    store = modulestore()
    for ccx in list_ccx:
        with store.bulk_operations(ccx.course_id):
            course = store.get_course(ccx.course_id)
        ccx_locator = CCXLocator.from_course_locator(ccx.course_id, unicode(ccx.id))
        reverse_add_master_course_staff_to_ccx(course, ccx_locator, ccx.display_name)
```

**Comment by giocalitri** [comment]
@pdpinch since we are talking about this, he may wants to make the migration self contained and not import the functions from `utils`

he also needs to import `modulestore`

**Comment by giocalitri** [comment]
@pdpinch I also would like to take a deep look at this PR once you give your thumbsup.

**Comment by giocalitri** [comment]
This PR should include the entire code currently in #207, given that most of that PR is about tests missing in the previous code and the actual changes in the rest APIs are trivial.

**Comment by giocalitri** [comment]
@pdpinch what do you think? ^^

**Comment by pdpinch** [comment]
Yeah. I don't think you'll pass coverage without the changes to test_utils.

**Comment by pdpinch** [comment]
You'll need to change your migration to 0003 and rebase this because https://github.com/edx/edx-platform/pull/11356 was merged.

**Comment by giocalitri** [comment]
I am not sure you picked both my comments from #207

**Comment by amir-qayyum-khan** [comment]
yes @giocalitri , one is squashed, now i am squashing second one.

**Comment by amir-qayyum-khan** [comment]
To run migration 

- ``python manage.py lms migrate ccx 0003 --settings=devstack1`` (This is to run forward migration)
- ``python manage.py lms migrate ccx 0002 --settings=devstack1`` (This is to run reverse migration)

**Comment by amir-qayyum-khan** [comment]
@giocalitri i rebased it with edx to see working with ur migration

**Comment by giocalitri** [comment]
I am going to review shortly

**Comment by giocalitri** [comment]
@amir-qayyum-khan 
1- please squash your commits
2- please make sure that only your commits appear in the diff of this PR

**Comment by giocalitri** [comment]
3- Please fix the description of your commit: the title should be short and add the rest in detail as body of the description. Plus you did not only fix the migration, since the entire reverted code should be there again.

**Comment by amir-qayyum-khan** [comment]
@giocalitri closed edX PR, 

FYI these tests are failing https://build.testeng.edx.org/job/edx-platform-test-subset/348737/console

**Comment by amir-qayyum-khan** [comment]
Done with fixing ccx/api failing tests,

**Comment by giocalitri** [comment]
this test is missing from your PR:
https://github.com/mitocw/edx-platform/blob/gdm_staff_in_ccx_%23205/lms/djangoapps/ccx/api/v0/tests/test_views.py#L465

are you sure you picked both commits from my PR?

File: lms/djangoapps/ccx/tests/test_utils.py [hunk 1]
```diff
@@ -93,3 +112,135 @@ def test_get_chapters(self):
...
+        # assert that staff and instructors of master course has staff and instructor roles on ccx
+        list_staff_master_course = list_with_level(self.course, 'staff')
+        list_instructor_master_course = list_with_level(self.course, 'instructor')
+
+        with ccx_course(self.ccx_locator) as course_ccx:
+            list_staff_ccx_course = list_with_level(course_ccx, 'staff')
+            self.assertNotEqual(len(list_staff_master_course), len(list_staff_ccx_course))
+
+            list_instructor_ccx_course = list_with_level(course_ccx, 'instructor')
+            self.assertNotEqual(len(list_instructor_ccx_course), len(list_instructor_master_course))
```
Comment (giocalitri) [inline lms/djangoapps/ccx/tests/test_utils.py]: what about other tests for the reverse? like calling it multiple times or tests for the display name?

I think I already mentioned that this function needs to be tested better

**Comment by giocalitri** [comment]
there are few tests to add and one to copy from the #207

**Comment by amir-qayyum-khan** [comment]
Added @giocalitri

File: lms/djangoapps/ccx/api/v0/tests/test_views.py [hunk 2]
```diff
@@ -590,22 +599,37 @@ def test_post_list(self):
         self.assertEqual(len(outbox), 1)
         self.assertIn(self.coach.email, outbox[0].recipients())  # pylint: disable=no-member
 
-    def test_post_list_duplicated_modules(self):
+    def test_post_list_staff_master_course_in_ccx(self):
```
Comment (giocalitri) [inline lms/djangoapps/ccx/api/v0/tests/test_views.py]: why this test is replacing `test_post_list_duplicated_modules`?

File: lms/djangoapps/ccx/tests/test_utils.py [hunk 3]
```diff
@@ -93,3 +112,179 @@ def test_get_chapters(self):
...
+
+            for user in list_staff_master_course:
+                self.assertNotIn(user, list_staff_ccx_course)
+            for user in list_instructor_master_course:
+                self.assertNotIn(user, list_instructor_ccx_course)
```
Comment (giocalitri) [inline lms/djangoapps/ccx/tests/test_utils.py]: what about the test for `send_email=False`

**Comment by giocalitri** [comment]
Can you please be sure you are thorough with the tests?

File: lms/djangoapps/ccx/api/v0/tests/test_views.py (hunk 2 above)
Comment (amir-qayyum-khan) [inline lms/djangoapps/ccx/api/v0/tests/test_views.py]: It seem I missed it during cherry-pick

**Comment by amir-qayyum-khan** [comment]
Added some more tests

**Comment by pdpinch** [comment]
will running the coverage tests locally pick up these testing gaps?

**Comment by amir-qayyum-khan** [comment]
yes , i should.

**Comment by amir-qayyum-khan** [comment]
@pdpinch this is coverage diff
> Diff: origin/master...HEAD, staged, and unstaged changes
-------------
cms/djangoapps/contentstore/views/course.py (0.0%): Missing lines 30,395-396,445
lms/djangoapps/ccx/api/v0/tests/test_views.py (96.3%): Missing lines 649
lms/djangoapps/ccx/api/v0/views.py (100%)
lms/djangoapps/ccx/tests/test_utils.py (100%)
lms/djangoapps/ccx/tests/test_views.py (100%)
lms/djangoapps/ccx/utils.py (100%)
lms/djangoapps/ccx/views.py (100%)
-------------
Total:   215 lines
Missing: 5 lines
Coverage: 97%
-------------

**Comment by pdpinch** [comment]
Ok, looking at the coverage, it looks like we don't have any tests for the changes to the studio dashboard (https://github.com/mitocw/edx-platform/pull/208/files#diff-968532c6bba75e1f61d1e1ddb51d7842). That should be pretty easy to add?

I forget what it means when there's no coverage in a test file (https://github.com/mitocw/edx-platform/pull/208/files#diff-456698559da70c26842813fa3a28c955R649). Does that mean that the line is never executed? It looks like we aren't testing instructors in addition to staff. Do we have instructors in our test cases?

**Comment by amir-qayyum-khan** [comment]
@pdpinch 
- The difficulty in [this path of code](https://github.com/mitocw/edx-platform/pull/208/files#diff-456698559da70c26842813fa3a28c955R649) is we can not create ccx in studio.

- yes we have, tests for instructors
https://github.com/mitocw/edx-platform/pull/208/files#diff-456698559da70c26842813fa3a28c955R638
https://github.com/mitocw/edx-platform/pull/208/files#diff-456698559da70c26842813fa3a28c955R638
 
> I forget what it means when there's no coverage in a test file (https://github.com/mitocw/edx-platform/pull/208/files#diff-456698559da70c26842813fa3a28c955R649). Does that mean that the line is never executed? It looks like we aren't testing instructors in addition to staff. Do we have instructors in our test cases?

I copied this part from @giocalitri #207, he can explain workflow better then me.

**Comment by pdpinch** [comment]
@amir-qayyum-khan can you step through the test code and find out why we're not executing https://github.com/mitocw/edx-platform/pull/208/files#diff-456698559da70c26842813fa3a28c955R649 -- I agree with you that it looks like it we should be hitting it. 

@giocalitri can I get a second opinion on the coverage gaps in studio? Is there no way we can test this? If that's the case, would it make sense to invert the test, i.e. instead of testing to see whether the course is an instance of a ccxLocator, we could test whether it's *not* an instance of a courseLocator. That seems less than ideal to me, but I just wanted to consider the options.

**Comment by giocalitri** [comment]
@pdpinch I never wrote unit tests for studio, but I find difficult to think that is impossible to create a CCX in such unit test. 
The codebase is the same and yes, maybe you do not have access at the LMS URLs, but I think you can still import the CCX model and create it manually.

@amir-qayyum-khan About the test at [this line](https://github.com/mitocw/edx-platform/pull/208/files#diff-456698559da70c26842813fa3a28c955R649) I do not see anything terribly mysterious that needs my explanation: if that line does not get hit during the test, it means there are no instructors in the master course and consequently in the CCX: to fix it add an instructor at the beginning of the test. 
Actually when I wrote the test there was 1 instructor and 1 staff in the master course, so I do not know why you do not have any, but again, it is pretty easy to solve.
In general I thought you took ownership of the code I wrote, so I do not understand why you keep referring me for issues with this code: if there are obscure parts please let me know so I can explain them better and we can change them to have clearer code.

**Comment by amir-qayyum-khan** [comment]
@giocalitri we cannot import code from lms into cms unless we put that code in common.

please see [common README](https://github.com/edx/edx-platform/blob/master/common/README.rst)

> common
------

This directory contains common code shared between LMS and CMS, such as Mako templates, CSS, and Coffescript.

**Comment by amir-qayyum-khan** [comment]
my suggestion is we need to some move code like models, utils to common.

@pdpinch @giocalitri

**Comment by pdpinch** [comment]
I think we're caught between a rock and hard place here. EdX would like CCX to be constrained to Lms. But they also want separation between CMS and Lms code (enabled by putting code in common).

I think we should leave this unresolved and open the PR on EdX and discuss with them in the review (I'm still trying to understand how this didn't come up earlier)

**Comment by amir-qayyum-khan** [comment]
@giocalitri are you ok with opening PR to edX?

cc @pdpinch

**Comment by pdpinch** [comment]
Did you figure out what's going on with https://github.com/mitocw/edx-platform/pull/208#issuecomment-193814392 ? 

----------------
I am mobile.

**Comment by amir-qayyum-khan** [comment]
yes , fixed in my last commit, now coverage is 

screen shot 2016-03-09 at 5 22 16 pm

**Comment by giocalitri** [comment]
@amir-qayyum-khan I did not know that document: I am sorry.

Anyway, have you tried to mock `modulestore().get_courses()` to return a `CCXLocator` instance?

**Comment by amir-qayyum-khan** [comment]
@giocalitri i am aleady working on mock approach, thanks

**Comment by amir-qayyum-khan** [comment]
@pdpinch @giocalitri done with adding tests for studio. (Test to check that ccx course filter out in studio dashboard).

File: cms/djangoapps/contentstore/tests/test_course_listing.py [hunk 4]
```diff
@@ -120,6 +122,39 @@ def test_get_course_list(self):
...
 
+    def test_get_course_list_when_ccx(self):
+        """
+        Assert that courses with CCXLocator are filter in course listing.
+        """
+        course_location = self.store.make_course_key('Org1', 'Course1', 'Run1')
+        self._create_course_with_access_groups(course_location, self.user)
+
+        # get courses through iterating all courses
+        courses_list, __ = _accessible_courses_list(self.request)
```
Comment (giocalitri) [inline cms/djangoapps/contentstore/tests/test_course_listing.py]: I believe the convention is to use a single `_`

Comment (amir-qayyum-khan) [inline cms/djangoapps/contentstore/tests/test_course_listing.py]: - @giocalitri i think we use single ``_`` for ``gettext`` and for unused variables we sue ``__``,   
- pylintrc take ``__`` and ``_`` as good name,

 [please see](https://github.com/edx/edx-platform/blob/master/pylintrc#L103)

Comment (giocalitri) [inline cms/djangoapps/contentstore/tests/test_course_listing.py]: yeah, i just remembered

**Comment by giocalitri** [comment]
this is good to go on edx :+1:

**Comment by giocalitri** [comment]
please squash before opening the PR against edx

**Comment by amir-qayyum-khan** [comment]
made PR upstream https://github.com/edx/edx-platform/pull/11787

**Pull request closed without merging**